### DISCUSSION TITLE
[Feature] CommunityToolkit.Extensions.DependencyInjection package

### DISCUSSION THREAD

**Discussion by Sergio0694** [CommunityToolkit/Labs-Windows#463] [Proposals]
We built a source generator (with analyzers) to help with dependency injection registration in the Microsoft Store, and we would like to open source this work into the Windows Community Toolkit Labs for Windows. This small library has the following advantages:
- Makes the service registration **completely trim/AOT-friendly**
- Provides built-time diagnostics on registration errors

This is how you can use the library:

```csharp
public sealed partial class App
{
    [Singleton(typeof(Logger), typeof(ILogger))]
    [Singleton(typeof(SomeService1), typeof(ISomeService1))]
    [Singleton(typeof(SomeService2), typeof(ISomeService2))]
    [Singleton(typeof(SomeService3), typeof(ISomeService3), typeof(IAnotherInterface))]
    [Singleton(typeof(StandaloneService))]
    [Transient(typeof(SomeViewModel))]
    // More services here...
    internal static partial void ConfigureServices(IServiceCollection services);
}
```

This will generate an implementation of that `ConfigureServices` method with registration calls for all declared services, and fully pre-generated reflection-free stubs to instantiate every service as well. That is, this will work just fine with AOT/trimming as well.

### API overview

The following attributes are included:

```csharp
namespace CommunityToolkit.Extensions.DependencyInjection;

[AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = false)]
[Conditional("SERVICES_CONFIGURATION_METADATA")]
public sealed class SingletonAttribute : Attribute
{
    public SingletonAttribute(Type implementationType, params Type[] serviceTypes);

    public Type ImplementationType { get; }
    public Type[] ServiceTypes { get; }
}

[AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = false)]
[Conditional("SERVICES_CONFIGURATION_METADATA")]
public sealed class TransientAttribute : Attribute
{
    public TransientAttribute(Type implementationType, params Type[] serviceTypes);

    public Type ImplementationType { get; }
    public Type[] ServiceTypes { get; }
}
```

We could of course also add `[Scoped]` if necessary.

**Comment by dongle-the-gadget** [comment]
What if we need to supply our own values?

**Reply by Sergio0694** [reply]
What do you mean "our own values", can you make an example?

**Reply by dongle-the-gadget** [reply]
I mean, providing our own implementation, for instance
`col.AddSingleton<InterfaceA, ImplementationA>(_ => new ImplementationA("something"));`

**Comment by AbdAlghaniAlbiek** [comment]
Then if you need inject these services inside ViewModels will be the same classic way (App.services.GetRequiredService<>()) OR it will be automatically using some "Attributes" like this:
```csharp
[SomeAttribute]
public partial class ViewModel 
{
    [CustomAttribute]
    private ISomeService1;

    [CustomAttribute]
    private ISomeService2;
}
```

**Comment by Mrxx99** [comment]
Wouldn't this make more sense in the MVVM toolkit instead of the windows toolkit, so it can be used not only with WinUI?

**Reply by michael-hawker** [reply]
I think that'd be where it would end up ultimately, we just don't have a similar '.NET Community Toolkit Labs' setup, so @Sergio0694 just wanted to use our infrastructure here for prototyping, feedback, and fine-tuning before figuring out where it goes. Right Sergio?

**Reply by Sergio0694** [reply]
I'm confused as to what the issue is. The package can already be used everywhere, it doesn't depend on WinUI at all (eg. we're using in on UWP in the Microsoft Store). It's just in the Labs for Windows repo for convenience, but the package just targets .NET Standard 2.0, so you can literally use it on whatever framework you want with no limitations 🙂

**Comment by MitchRazga** [comment]
I would be very interested in this experiment. Sergio and the team have an excellent track record for delivering high-quality and performant source generators. However, I am curious to know how this would distinguish itself or offer any advantages compared to other popular solutions, such as [Jab](https://github.com/pakrym/jab) and [StrongInject](https://github.com/YairHalberstadt/stronginject)

**Comment by Arlodotexe** [comment]
This experiment is now available in Labs, [here](https://github.com/CommunityToolkit/Labs-Windows/tree/main/components/Extensions.DependencyInjection).

**Comment by dongle-the-gadget** [comment]
@Sergio0694 I installed version 0.1.230830 and it seems that the generator would create code that passes in the `IServiceCollection` into `ServiceProviderServiceExtensions.GetRequiredService` as an argument, which doesn't work (as the method only accepts an `IServiceProvider`)

Note that this bug doesn't reproduce if the `IServiceCollection` argument is named `services`.

### Minimal sample to trigger this bug
```cs
using CommunityToolkit.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection;

namespace TKDIBugRepo;

static partial class Program
{
    static void Main()
    {

    }

    [Singleton(typeof(Foo))]
    [Singleton(typeof(Test))]
    static partial void ConfigureServices(IServiceCollection collection);
}

public class Foo
{ }

public class Test
{
    private readonly Foo _foo;

    public Test(Foo foo)
        => _foo = foo;
}
```

### Generated code
```cs
// <auto-generated/>
#pragma warning disable
#nullable enable
 
namespace TKDIBugRepo
{
    /// <inheritdoc/>
    partial class Program
    {
        /// <inheritdoc/>
        [global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Extensions.DependencyInjection.SourceGenerators.ServiceProviderGenerator", "1.0.0.0")]
        [global::System.Diagnostics.DebuggerNonUserCode]
        [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
        static partial void ConfigureServices(global::Microsoft.Extensions.DependencyInjection.IServiceCollection collection)
        {
            global::Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton(collection, typeof(global::TKDIBugRepo.Foo), new global::TKDIBugRepo.Foo());
            global::Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton(collection, typeof(global::TKDIBugRepo.Test), static services => new global::TKDIBugRepo.Test(
                global::Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService<global::TKDIBugRepo.Foo>(collection)));
        }
    }
}
```

### Suggestion for fix
Replace `info.Method.ServiceCollectionParameterName` with `services`?

https://github.com/CommunityToolkit/Labs-Windows/blob/dff647f50adad49cf049e924e3494a4381c6a94f/components/Extensions.DependencyInjection/CommunityToolkit.Extensions.DependencyInjection.SourceGenerators/ServiceProviderGenerator.Execute.cs#L195

**Comment by antoniovalentini** [comment]
Hey @Sergio0694, the new package is not in the official feed yet, right? Any ETA for when (or if) this will happen?
Let us know if we can help as a community in any way. Thanks for your hard work 🙏

**Comment by shmuelie** [comment]
Requesting support for registering more than one implementation of an interface, something like

```csharp
public class Something(IEnumerable<IService> service) { }

[SingletonGroup(typeof(Impl1), typeof(IService))]
[SingletonGroup(typeof(Impl2), typeof(IService))]
[Singleton(typeof(Something))]
internal static partial void ConfigureServices(IServiceCollection services);
```

**Reply by michael-hawker** [reply]
Yeah, I'm thinking too just of like `[SingletonGroup(typeof(IService))]` where it's a group of just anything with that interface too.

**Reply by michael-hawker** [reply]
Was just looking for at least @shmuelie's case again. @Sergio0694 are there any tests in the store to be brought over? It'd make it a bit easier to setup a new scenario for this to see if a new attribute is needed or not. I figured I'd try it out quick, but then get the TKEXDI0004 message about the service type interface already being registered to the target service collection.

**Comment by shmuelie** [comment]
Requesting support for [Keyed Services](https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection#keyed-services):

```csharp

public class SomethingA([FromKeyedServices("key1")]IService service) { }
public class SomethingB(IEnumerable<IService> service) { }

[Singleton("key1", typeof(Impl1), typeof(IService))]
[Singleton("key2", typeof(Impl2), typeof(IService))]
[Singleton(typeof(SomethingA))]
[Singleton(typeof(SomethingB))]
internal static partial void ConfigureServices(IServiceCollection services);
```

Asking for the services as enumerable without a key should use [`KeyedService.AnyKey`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.keyedservice.anykey)
